### PR TITLE
Parse CsvSource.value[] elements  separately

### PR DESCRIPTION
**Pull request by mpkorstanje**
In JUnit v5, when using `CsvSource.value[]` the expectation is that user declares individual CSV records. When declaring multiple records either through a text block, or explicit new-line, any records after the first were ignored.

However in v6 we switched parsers to FastCSV in https://github.com/junit-team/junit-framework/pull/4606. This changed the implementation. Instead of treating each `value` as a single record, all the `value[]` elements are joined with `\n` and then treated as if they were a single CSV document.

This has the benefit that multiple records can now be declared in a single string, but it is now also possible that unclosed qoutes from one record can bleed into the next record. We don't consider the latter to be desirable.

For example:

```
@CsvSource(
  "a, 'b", 
  "c', d"
)
```

Should provide two records, containing two fields each. Instead we find fields `a`, `b\nc` and `d`. This should be fields `a`, `b`, `c'` and `d`. Note: The parser leniently closes the unclosed quote for b.

By treating each `value[]` element as if it was an dependent document we prevent quotes from bleeding over while still allowing each `value[]` element to contain multiple records

Contributes to: #5017

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit-framework/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [ ] There are no TODOs left in the code
- [ ] Method [preconditions](https://docs.junit.org/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [ ] [Coding conventions](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [ ] Change is covered by [automated tests](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [ ] Public API has [Javadoc](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [ ] Change is documented in the [User Guide](https://docs.junit.org/snapshot/) and [Release Notes](https://docs.junit.org/snapshot/release-notes.html)

### CODE REVIEW

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvReaderFactory.java [hunk 1]
```diff
@@ -88,7 +89,8 @@ static CsvReader<? extends CsvRecord> createReaderFor(CsvSource csvSource, Strin
 				Set.of(csvSource.nullValues()),
 				csvSource.ignoreLeadingAndTrailingWhitespace(),
 				csvSource.maxCharsPerColumn(),
-				csvSource.useHeadersInDisplayName()
+				// For CsvSource.value we manually process the header.
+				hasTextBlock && csvSource.useHeadersInDisplayName()
```
Comment (mpkorstanje) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvReaderFactory.java]: Implicit repeated condition in provideArgumentsFromValueArray. De-duplicate and pass as parameter.

Comment (mpkorstanje) [reply junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvReaderFactory.java]: Maybe wrap `CsvSource` and `CsvFileSource` so we have all this logic in one place.

**Review by marcphilipp — COMMENTED** [review]
Looks very reasonable! Is there anything you'd like feedback on?

**Comment by mpkorstanje** [comment]
> Looks very reasonable! Is there anything you'd like feedback on?

No. Biggest point of discussion would be what to decide what, if anything, to do after this.

This PR is far from done though. At the moment there are two effective configurations spread out over three different code paths. I'd like to reduce the rube-goldberg-esq nature somewhat.

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvSource.java [hunk 2]
```diff
@@ -122,20 +118,54 @@
 	 *     // ...
 	 * }</pre>
 	 *
+	 * <h4>Text Blocks</h4>
```
Comment (mpkorstanje) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvSource.java:121]: The implementation for `CsvSource.value` supports whole CSV documents while the documentation pretends it only supports records. Trying to fix that pretense makes the explanation of the basic use case unreadable, so I've duplicated most of the the `textBlock` documentation here as pretend "different" usecase.

I'm open to better ideas.

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvReaderConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+
+import de.siegmar.fastcsv.reader.CommentStrategy;
+
+import org.junit.platform.commons.util.Preconditions;
+
+/**
+ * Wrapper for {@link CsvSource} with {@code value}, {@link CsvSource} with a
+ * {@code textBlock} and {@link CsvFileSource}.
+ */
+abstract class CsvReaderConfiguration {
```
Comment (mpkorstanje) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvReaderConfiguration.java]: This class is rather verbose, but it also makes it explicit that we have 3 similar but different use-cases (`CsvSource.value`, `CsvSource.textBlock`, `CsvFileSource`) and that `CsvSource.textBlock` are `CsvFileSource` identical.

I prefer it over having two different but similar methods to create `CsvParsers`, but that is just a preference.

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvSource.java (hunk 2 above)
Comment (marcphilipp) [reply junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvSource.java:121]: I think that's fine and I don't have a better idea.

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvReaderConfiguration.java (hunk 3 above)
Comment (marcphilipp) [reply junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvReaderConfiguration.java]: Probably a nit-pick, but I wonder if it would be clearer/shorter if this class was a `record` which would avoid the delegation pattern. I'll let you decide, though.

**Review by marcphilipp — APPROVED** [review]

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvReaderConfiguration.java (hunk 3 above)
Comment (mpkorstanje) [reply junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvReaderConfiguration.java]: It does look much better.
